### PR TITLE
PIL-758 Contact address page

### PR DESCRIPTION
**Pull request by javed-iqbal1**

### CODE REVIEW

File: app/views/rfm/RfmContactAddressView.scala.html [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+@layout(pageTitle = title(form, messages("rfmContactAddress.title"))) {
+
+    @formHelper(action = controllers.rfm.routes.RfmContactAddressController.onSubmit(mode)) {
+
+        @if(form.errors.nonEmpty) {
+            @govukErrorSummary(ErrorSummaryViewModel(form))
+        }
+        @sectionHeader(messages("rfmContactAddress.heading.caption"))
+        <h1 class="govuk-heading-l long-word">@messages("rfmContactAddress.heading")</h1>
```
Comment (AKKUM) [inline app/views/rfm/RfmContactAddressView.scala.html:58]: Can we put heading gds component?

File: test/controllers/rfm/RfmContactAddressControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+              ("countryCode", "GB")
+            )
+
+        val result = route(application, request).value
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result).value mustEqual controllers.routes.UnderConstructionController.onPageLoad.url
+
+      }
+    }
+
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactAddressControllerSpec.scala:128]: May need to check post code is mandatory, if country is GB.

File: test/controllers/rfm/RfmContactByTelephoneControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        contentAsString(result) mustEqual view(formProvider, NormalMode, "sad")(
+          request,
+          appConfig(application),
+          messages(application)
+        ).toString
+
+      }
+    }
+
+    "send  to recovery page if no name is provided" in {
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactByTelephoneControllerSpec.scala:63]: Do we have to send to recovery page when answer is no?

File: test/controllers/rfm/RfmContactByTelephoneControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            .withFormUrlEncodedBody(("value", "true"))
+
+        val result = route(application, request).value
+
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result).value mustEqual controllers.rfm.routes.RfmCapturePrimaryTelephoneController.onPageLoad(NormalMode).url
+      }
+    }
+
+    "Show error when no option is selected " in {
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactByTelephoneControllerSpec.scala:142]: Where are having no option selected here?

File: app/forms/RfmCaptureTelephoneDetailsFormProvider.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package forms
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+import javax.inject.Inject
+
+class RfmCaptureTelephoneDetailsFormProvider @Inject() extends Mappings {
```
Comment (AKKUM) [inline app/forms/RfmCaptureTelephoneDetailsFormProvider.scala:24]: Please cover the unit test for this.

File: app/forms/RfmContactAddressFormProvider.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+package forms
+
+import forms.mappings.{AddressMappings, Mappings}
+import models.UKAddress
+import play.api.data.Form
+import play.api.data.Forms.{mapping, optional}
+
+import javax.inject.Inject
+
+class RfmContactAddressFormProvider @Inject() extends Mappings with AddressMappings {
```
Comment (AKKUM) [inline app/forms/RfmContactAddressFormProvider.scala:26]: Please cover the unit tests for this.

File: app/forms/RfmContactByTelephoneFormProvider.scala [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package forms
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+import javax.inject.Inject
+
+class RfmContactByTelephoneFormProvider @Inject() extends Mappings {
```
Comment (AKKUM) [inline app/forms/RfmContactByTelephoneFormProvider.scala:24]: Same here unit test for this

File: app/forms/RfmPrimaryContactEmailFormProvider.scala [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+package forms
+
+import forms.mappings.Mappings
+import mapping.Constants
+import play.api.data.Form
+
+import javax.inject.Inject
+
+class RfmPrimaryContactEmailFormProvider @Inject() extends Mappings {
```
Comment (AKKUM) [inline app/forms/RfmPrimaryContactEmailFormProvider.scala:25]: same test

File: app/pages/RfmCapturePrimaryTelephonePage.scala [hunk 9]
```diff
@@ -18,9 +18,10 @@ package pages
 
 import play.api.libs.json.JsPath
 
-case object RfmPrimaryNameRegistrationPage extends QuestionPage[String] {
+case object RfmCapturePrimaryTelephonePage extends QuestionPage[String] {
```
Comment (AKKUM) [inline app/pages/RfmCapturePrimaryTelephonePage.scala:21]: Can we have page spec aswell.

File: app/pages/RfmContactAddressPage.scala [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pages
+
+import models.UKAddress
+import play.api.libs.json.JsPath
+
+case object RfmContactAddressPage extends QuestionPage[UKAddress] {
```
Comment (AKKUM) [inline app/pages/RfmContactAddressPage.scala:22]: Can we page specs.

File: app/pages/RfmContactByTelephonePage.scala [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package pages
+
+import models.UserAnswers
+import play.api.libs.json.JsPath
+
+import scala.util.Try
+
+case object RfmContactByTelephonePage extends QuestionPage[Boolean] {
```
Comment (AKKUM) [inline app/pages/RfmContactByTelephonePage.scala:24]: Page specs for this as well.

File: app/pages/RfmPrimaryContactEmailPage.scala [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pages
+
+import play.api.libs.json.JsPath
+
+case object RfmPrimaryContactEmailPage extends QuestionPage[String] {
```
Comment (AKKUM) [inline app/pages/RfmPrimaryContactEmailPage.scala:21]: Page spec here as well.

File: app/pages/RfmPrimaryContactNamePage.scala [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pages
+
+import play.api.libs.json.JsPath
+
+case object RfmPrimaryContactNamePage extends QuestionPage[String] {
```
Comment (AKKUM) [inline app/pages/RfmPrimaryContactNamePage.scala:21]: page spec here.

File: test/controllers/rfm/RfmCapturePrimaryTelephoneControllerSpec.scala [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    }
+    "return bad request if wrong data is inputted" in {
+      val ua          = emptyUserAnswers.set(RfmPrimaryContactNamePage, "sad").success.value
+      val application = applicationBuilder(userAnswers = Some(ua)).build()
+      running(application) {
+        val request =
+          FakeRequest(POST, controllers.rfm.routes.RfmCapturePrimaryTelephoneController.onSubmit(NormalMode).url)
+            .withFormUrlEncodedBody("value" -> "adsasd")
+        val result = route(application, request).value
+        status(result) mustEqual BAD_REQUEST
```
Comment (AKKUM) [inline test/controllers/rfm/RfmCapturePrimaryTelephoneControllerSpec.scala:131]: We need to test content here please test contentAsString(result)

File: test/controllers/rfm/RfmContactAddressControllerSpec.scala [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    "return OK and the correct view for a GET with no previous answer" in {
+      val ua = emptyUserAnswers.set(RfmPrimaryContactNamePage, "Name").success.value
+      val application = applicationBuilder(userAnswers = Some(ua))
+        .build()
+
+      running(application) {
+        val request = FakeRequest(GET, controllers.rfm.routes.RfmContactAddressController.onPageLoad(NormalMode).url)
+
+        val result = route(application, request).value
+        status(result) mustEqual OK
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactAddressControllerSpec.scala:50]: Please test the content. contentAsString(result)

File: test/controllers/rfm/RfmContactAddressControllerSpec.scala [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+      running(application) {
+        val request = FakeRequest(GET, controllers.rfm.routes.RfmContactAddressController.onPageLoad(NormalMode).url)
+        val view    = application.injector.instanceOf[RfmContactAddressView]
+        val result  = route(application, request).value
+        status(result) mustEqual OK
+        contentAsString(result) must include(
+          "What address do you want to use as the filing members contact address?"
+        )
+        contentAsString(result) must include(
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactAddressControllerSpec.scala:96]: Can you please test the whole content.

File: test/controllers/rfm/RfmContactAddressControllerSpec.scala [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+              ("addressLine2", "Drive"),
+              ("addressLine3", "Newcastle"),
+              ("addressLine4", "North east"),
+              ("postalCode", "hhhhhhhhhhhh"),
+              ("countryCode", "GB")
+            )
+
+        val result = route(application, request).value
+        status(result) mustEqual BAD_REQUEST
+
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactAddressControllerSpec.scala:150]: Test the content contentAsString(result)

File: test/controllers/rfm/RfmContactAddressControllerSpec.scala [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+              ("addressLine3", "Newcastle"),
+              ("addressLine4", "North east"),
+              ("postalCode", "ne5 2th"),
+              ("countryCode", "GB")
+            )
+
+        val result = route(application, request).value
+
+        status(result) mustEqual BAD_REQUEST
+      }
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactAddressControllerSpec.scala:180]: Same here contentAsString(result)

File: test/controllers/rfm/RfmContactByTelephoneControllerSpec.scala [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+        .build()
+
+      running(application) {
+        when(mockUserAnswersConnectors.save(any(), any())(any())).thenReturn(Future(Json.toJson(Json.obj())))
+        val request =
+          FakeRequest(POST, controllers.rfm.routes.RfmContactByTelephoneController.onSubmit(NormalMode).url)
+
+        val result = route(application, request).value
+
+        status(result) mustEqual BAD_REQUEST
```
Comment (AKKUM) [inline test/controllers/rfm/RfmContactByTelephoneControllerSpec.scala:159]: Same here contentAsString(result)

**Review by jassalrichy — APPROVED** [review]

**Review by jassalrichy — DISMISSED** [review]

**Review by jassalrichy — DISMISSED** [review]

**Review by jassalrichy — APPROVED** [review]

**Review by jassalrichy — DISMISSED** [review]

**Review by jassalrichy — APPROVED** [review]

**Pull request merged**